nx/lib/nx/random.ex

```diff
@@ -601,6 +601,245 @@ defmodule Nx.Random do
+      iex> result
+      #Nx.Tensor<
+        s64[10]
+        [3, 1, 2, 5, 4, 0]
+      >
+  """
+  deftransform choice(key, tensor, p), do: choice_probabilities(key, tensor, p)
+  deftransform choice(key, tensor, p, opts), do: choice_probabilities(key, tensor, p, opts)
+
+  deftransformp validate_choice_opts(tensor, opts) do
+    opts = Keyword.validate!(opts, [:samples, :axis, replace: true])
+
+    {axis, tensor_shape} =
+      case opts[:axis] do
+        nil ->
+          {0, {Tuple.product(tensor.shape)}}
+
+        axis ->
+          {Nx.Shape.normalize_axis(tensor.shape, axis, tensor.names), tensor.shape}
+      end
+
+    if Nx.rank(tensor) < 1 do
+      raise ArgumentError, "tensor must have rank 1 or greater"
+    end
+
+    n_draws = opts[:samples]
+
+    if n_draws < 1 do
+      raise "must take at least one sample, got samples=#{n_draws}"
+    end
+
+    n_inputs =
+      case opts[:axis] do
+        nil -> Nx.size(tensor)
+        _ -> Nx.axis_size(tensor, axis)
+      end
+
+    replace = opts[:replace]
+
+    if not replace and n_draws > n_inputs do
+      raise ArgumentError, "cannot take more samples than the input size when replace: false"
+    end
+
+    {tensor_shape, n_inputs, n_draws, axis, replace}
+  end
+
+  defnp choice_uniform(key, tensor, opts) do
+    {tensor_shape, n_inputs, n_draws, axis, replace} = validate_choice_opts(tensor, opts)
+    tensor = Nx.reshape(tensor, tensor_shape)
+
+    if replace do
+      {idx, key} = randint(key, 0, n_inputs, shape: {n_draws})
+      result = Nx.take(tensor, idx, axis: axis)
+      {result, key}
+    else
+      {shuffled, key} = shuffle(key, tensor, axis: axis)
+      result = Nx.slice_along_axis(shuffled, 0, n_draws, axis: axis)
+      {result, key}
+    end
+  end
+
+  defnp choice_probabilities(key, tensor, p, opts \\ []) do
+    {tensor_shape, n_inputs, n_draws, axis, replace} = validate_choice_opts(tensor, opts)
+    tensor = Nx.reshape(tensor, tensor_shape)
+
+    case {Nx.size(p), Nx.axis_size(tensor, axis)} do
+      {n, n} ->
+        :ok
+
+      _ ->
+        raise ArgumentError, "input and probabilities tensors must have the same shape"
+    end
+
+    if replace do
+      p_cumulative = Nx.cumulative_sum(p)
+      {uniform, key} = uniform(key, shape: {n_draws}, type: Nx.type(p_cumulative))
+      r = p_cumulative[-1] * (1 - uniform)
+
+      # naïve implementation of jax.numpy.searchsorted
+      p_cumulative = Nx.new_axis(p_cumulative, 0)
+      r = Nx.new_axis(r, 1)
+      idx = Nx.argmin(p_cumulative <= r, tie_break: :low, axis: 1)
+
+      result = Nx.take(tensor, idx, axis: axis)
+      {result, key}
+    else
+      {g, k} = gumbel(key, shape: {n_inputs}, type: Nx.type(p))
+      g = -g - Nx.log(p)
+      idx = g |> Nx.argsort() |> Nx.slice_along_axis(0, n_draws, axis: 0)
+
+      result = Nx.take(tensor, idx, axis: axis)
+      {result, k}
+    end
+  end
+
+  @doc """
+  Sample Gumbel random values with given shape and float dtype.
```

Comment:
Should I expose this function, or should we keep it hidden until someone actually needs it?

If it should be exposed, I need to add doctests

Reply:
Sounds fine by me. Should we also add `_split` versions to all of these functions?

Reply:
What does `_split` do?

Reply:
@josevalim Actually,  this depends on `shuffle`,  which `split`s the key underneath a few times. Does it still make sense to have `choice_split`?

Reply:
Ah, good point. Try to have a `_split` if we can, but if you cannot, no problem!

Reply:
I don't see a point of having `choice_split` because of `shuffle`, but we can have `gumbel_split`